core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMNode.kt

```diff
@@ -61,4 +91,44 @@ data class STDCMNode(
             waypointIndex
         )
     }
+
+    /**
+     * Returns the relative time difference between the real arrival time at node and the planned
+     * arrival time at node, taking into account the tolerance on either side. It takes the lowest
+     * value between the relative time diff at the current time and the relative time diff at the
+     * maximum time at which the train can pass on the node.
+     */
+    fun getRelativeTimeDiff(currentTotalAddedDelay: Double = totalPrevAddedDelay): Double? {
+        if (plannedTimingData != null) {
+            val timeDiff =
+                getRealTime(currentTotalAddedDelay) - plannedTimingData.arrivalTime.seconds
+            val relativeTimeDiff = getLeftOrRightRelativeTimeDiff(timeDiff)
+            val maxTimeDiff =
+                min(
+                    getRealTime(maximumAddedDelay) - plannedTimingData.arrivalTime.seconds,
+                    plannedTimingData.arrivalTimeToleranceAfter.seconds
+                )
+            val relativeMaxTimeDiff = getLeftOrRightRelativeTimeDiff(maxTimeDiff)
+            val minRelativeTimeDiff = min(relativeTimeDiff, relativeMaxTimeDiff)
+            return minRelativeTimeDiff
+        }
+        return null
+    }
+
+    private fun getLeftOrRightRelativeTimeDiff(timeDiff: Double): Double {
+        return when {
+            timeDiff < 0 -> abs(timeDiff / plannedTimingData!!.arrivalTimeToleranceBefore.seconds)
+            timeDiff > 0 -> timeDiff / plannedTimingData!!.arrivalTimeToleranceAfter.seconds
+            else -> timeDiff
+        }
+    }
+
+    /**
+     * Takes into account the real current departure time shift to return the real current time at
+     * which the train arrives at this node.
+     */
+    private fun getRealTime(currentTotalAddedDelay: Double): Double {
```

Comment:
We could add a comment on `node.time` to explain that it's not *necessarily* the actual time, and that this method should be used to get it

Reply:
Done in https://github.com/OpenRailAssociation/osrd/pull/8152/commits/d73732b06b00a061c175bb5d2a0db657753bf346.